pkg/apis/core/validation/validation.go

```diff
@@ -4338,6 +4338,10 @@ func ValidateService(service *core.Service, allowAppProtocol bool) field.ErrorLi
 		}
 	}
 
+	if service.Spec.AllocateLoadBalancerNodePorts != nil && service.Spec.Type != core.ServiceTypeLoadBalancer {
```

Comment:
We don't usually do gates in validation, but I think it's appropriate to add a gated check here that if the gate is enabled a value is required.  We only allow nil if the gate is disabled (see other comments on using defaults.go)

Reply:
I would like to keep `nil` as default. Main reasons;

* On beta when default is true for the feature-gw no new attribute will pop up on a `kubectl get svc`. I guess like 95% of users will not set this attribute and don't want to see it.
* A web-hook was proposed as the solution for automatic setting of this parameter. If the default is `true` the hook can't tell if this is a default (should be set to false) or explicitly set to `true` by the used (should be left as-is).

Or will the web-hook be called before "default.go"? And will values with default values also be invisible on a `kubectl get svc`?

If so, default=true is fine. But then, is there a reson for the attribute to be a pointer? I have read that this is recommended for API attributes someplace.

Reply:
> no new attribute will pop up on a kubectl get svc. I guess like 95% of users will not set this attribute and don't want to see it.

We don't want to hide fields that are actually set.  Whether the user knows it or not, we are setting a value.  Most of the time, when we have done this we have regretted it.

> A web-hook was proposed as the solution for automatic setting of this parameter. If the default is true the hook can't tell if this is a default (should be set to false) or explicitly set to true by the used (should be left as-is).

This is a general problem and this is the 3rd or 4th time it has come up this month.  https://github.com/kubernetes/kubernetes/pull/95269 for another.  @apelisse @liggitt we need to either solve this generically or we decide that we are not handling it at all.

@uablrek the problem with leaving it nil is that we lose information.  At defaults.go time we know which API version was used to create the resource and we can apply version-specific default values.  In v1 we default to `true` but maybe in v2 we don't.  If we leave it nil we have no idea.  We can set this much later in REST, but before we really open that can of worms, let's discuss more.

I think it's best to do it the "normal" way (defaults.go) and let the defaulting vs webhooks debate be had in parallel. :)

Making it a pointer makes it possible to know AT ALL (defaults.go or later) whether the user specified `false` or didn't specify anything at all (Go's zero value).  `bool` fields that default to `true` (generally: any scalar type whose default != Go's zero value AND whose zero-value is in-bounds) are the one case where apointer is absolutely required.

Reply:
Ok, I'll update (and re-base). Should the value remain `nil` if the feature-gate is not set?

Reply:
> Should the value remain nil if the feature-gate is not set?

Yes, it should be `nil` when the feature gate is disabled. And even when the feature gate is enabled, I think `allocateLoadBalancerNodePorts` should be nil when the type is **not** LoadBalancer.

Reply:
Good point Andrew.

Reply:
> it's appropriate to add a gated check here that if the gate is enabled a value is required. We only allow nil if the gate is disabled (see other comments on using defaults.go)

A non-nil is ensured in (the updated) defaults.go if the gate is enabled. I did not alter the check.

I can add a check that the value is _not_ set unless the gate is enabled, but it seems unnecessary.